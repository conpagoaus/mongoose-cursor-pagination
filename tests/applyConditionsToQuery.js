const mongoose = require('mongoose')
const { applyConditionsToQuery } = require('../lib/utils')

const Comment = mongoose.model('comment')

const singleKeyCursorCondition = [{
   _id: { $lt: '5b06b90b42a0b29ba10f20c2' }
}]

const multiKeyCursorConditions = [
   { 'author.lastName': { $gt: 'Doe' } },
   {
      'author.lastName': 'Doe',
      _id: { $lt: '5b06b90b42a0b29ba10f20c2' }
   }
]

describe('utils#applyConditionsToQuery', () => {
   it('Works with single cursor keys', () => {
      let query = Comment.find({ body: 'Test' })
      query = applyConditionsToQuery(singleKeyCursorCondition, query)
      expect(query.getQuery()).toMatchSnapshot()
   })

   it('Works with multiple cursor keys with no other $or query conditions', () => {
      let query = Comment.find({ body: 'Test' })
      query = applyConditionsToQuery(multiKeyCursorConditions, query)
      expect(query.getQuery()).toMatchSnapshot()
   })

   it('Works with multiple cursor keys when there is an $or condition in the query', async () => {
      let query = Comment.find({
         $or: [
            { body: 'Test' },
            { 'author.firstName': 'Oreo' }
         ]
      })

      query = applyConditionsToQuery(multiKeyCursorConditions, query)
      expect(query.getQuery()).toMatchSnapshot()
      await query.exec()
   })

   it('Works with multiple cursor keys when there is an $or and $and condition in the query', async () => {
      let query = Comment.find({
         $and: [
            { body: 'body is in an and' },
            { 'author.firstName': 'Andy' }
         ],
         $or: [
            { body: 'body is in an or' },
            { 'author.firstName': 'Oreo' }
         ]
      })

      query = applyConditionsToQuery(multiKeyCursorConditions, query)
      expect(query.getQuery()).toMatchSnapshot()
      await query.exec()
   })
})
