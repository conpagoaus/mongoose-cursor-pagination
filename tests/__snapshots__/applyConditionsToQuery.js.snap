// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils#applyConditionsToQuery Works with multiple cursor keys when there is an $or and $and condition in the query 1`] = `
Object {
  "$and": Array [
    Object {
      "body": "body is in an and",
    },
    Object {
      "author.firstName": "Andy",
    },
    Object {
      "$or": Array [
        Object {
          "body": "body is in an or",
        },
        Object {
          "author.firstName": "Oreo",
        },
      ],
    },
    Object {
      "$or": Array [
        Object {
          "author.lastName": Object {
            "$gt": "Doe",
          },
        },
        Object {
          "_id": Object {
            "$lt": "5b06b90b42a0b29ba10f20c2",
          },
          "author.lastName": "Doe",
        },
      ],
    },
  ],
}
`;

exports[`utils#applyConditionsToQuery Works with multiple cursor keys when there is an $or condition in the query 1`] = `
Object {
  "$and": Array [
    Object {
      "$or": Array [
        Object {
          "body": "Test",
        },
        Object {
          "author.firstName": "Oreo",
        },
      ],
    },
    Object {
      "$or": Array [
        Object {
          "author.lastName": Object {
            "$gt": "Doe",
          },
        },
        Object {
          "_id": Object {
            "$lt": "5b06b90b42a0b29ba10f20c2",
          },
          "author.lastName": "Doe",
        },
      ],
    },
  ],
}
`;

exports[`utils#applyConditionsToQuery Works with multiple cursor keys with no other $or query conditions 1`] = `
Object {
  "$or": Array [
    Object {
      "author.lastName": Object {
        "$gt": "Doe",
      },
    },
    Object {
      "_id": Object {
        "$lt": "5b06b90b42a0b29ba10f20c2",
      },
      "author.lastName": "Doe",
    },
  ],
  "body": "Test",
}
`;

exports[`utils#applyConditionsToQuery Works with single cursor keys 1`] = `
Object {
  "_id": Object {
    "$lt": "5b06b90b42a0b29ba10f20c2",
  },
  "body": "Test",
}
`;
